paddle/fluid/operators/math/functors.h

```diff
@@ -0,0 +1,383 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+#include "paddle/fluid/framework/tensor.h"
+#include "paddle/fluid/framework/tensor_util.h"
+#include "paddle/fluid/operators/elementwise_op_function.h"
+#include "paddle/fluid/platform/device_context.h"
+
+namespace paddle {
+namespace operators {
+namespace math {
+
+// AddFunctor
+template <typename T>
+struct AddFunctor {
+  // out = x + y;
+  inline HOSTDEVICE T operator()(const T x, const T y) const { return x + y; }
+};
+
+template <typename T>
+struct AddGradFunctor {
+  inline HOSTDEVICE T operator()(const T x, const T y) const { return 1; }
+
+  inline HOSTDEVICE T operator()(const T x, const T y, const T out) const {
+    return 1;
+  }
+};
+
+// ScaleFunctor
+template <typename T>
+struct ScaleFunctor {
+  explicit ScaleFunctor(const T coeff) : coeff_(coeff) {}
+
+  inline HOSTDEVICE T operator()(const T ele) const { return ele * coeff_; }
+
+ private:
+  T coeff_;
+};
+
+template <typename T>
+struct ScaleGradFunctor {
+  explicit ScaleGradFunctor(T coeff) : coeff_(coeff) {}
+
+  inline HOSTDEVICE T operator()(const T x) const { return coeff_; }
+
+  inline HOSTDEVICE T operator()(const T x, const T out) const {
+    return coeff_;
+  }
+
+ private:
+  T coeff_;
+};
+
+// ReluFunctor
+template <typename T>
+struct ReluFunctor {
+  inline HOSTDEVICE T operator()(const T x) const { return x * (x > 0); }
+};
+
+template <typename T>
+struct ReluGradFunctor {
+  inline HOSTDEVICE T operator()(const T x) const { return x > 0 ? 1 : 0; }
+
+  inline HOSTDEVICE T operator()(const T x, const T out) const {
+    return x > 0 ? 1 : 0;
+  }
+};
+
```

Comment:
The classes below this line should be the private class for `fused_element_wise_op`. Should not in `math`.

Reply:
Done